File: go/102.go
/**
题目概述:

给你一个二叉树，请你返回其按 层序遍历 得到的节点值。 （即逐层地，从左到右访问所有节点）。


*/
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
//递归
var res [][]int
func levelOrder(root *TreeNode) [][]int {
       if root ==nil{
       	   return nil
	   }
	  res=make([][]int,0)
	  dfs(root,0)
	  return res

}

func dfs(root *TreeNode, level int)  {
	if root==nil{
		return
	}
	if level==len(res){
		res=append(res,[]int{})
	}
	res[level]=append(res[level],root.Val)
    dfs(root.Left,level+1)
    dfs(root.Right,level+1)
}

func levelOrder(root *TreeNode) [][]int {
    result:=[][]int{}
    if root ==nil{
    	return result
	}
    queue:=[]*TreeNode{root}
    for level:=0;0<len(queue);level++{ // 层级遍历
      result = append(result, []int{})
      next:=[]*TreeNode{}
      for j:=0;j<len(queue);j++{ // 层级元素遍历
      	node:=queue[j]
      	result[level]=append(result[level],node.Val)
      	if node.Left !=nil{
      		next=append(next,node.Left)
		}
      	if node.Right !=nil{
      		next=append(next,node.Right)
		}
	  }
      queue=next
	}
  return result
}

File: php/102.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

   private $res=[];
    /**
     * @param TreeNode $root
     * @return Integer[][]
     */
    function levelOrder($root) {
        // 递归
        $this->dfs($root,0);
        return  $this->res;
    }

    function dfs($root,$level){
        if($root==null){
            return $this->res;
        }
        $this->res[$level][]=$root->val;
        $this->dfs($root->left,$level+1);
        $this->dfs($root->right,$level+1);

    }

      // 迭代
      function levelOrder($root) {
         $result=[];
         if($root==null){
           return $result;
         }
         $queue=[$root];

         for($level=0;0<count($queue);$level++){
            $next=[];
            for($j=0;$j<count($queue);$j++){
               $node=$queue[$j];
               $result[$level][]=$node->val;
               if($node->left !=null){
                 $next[]=$node->left;
               }
               if($node->right !=null){
                $next[]=$node->right;
               }
            }
            $queue=$next;
         }
         return $result;
      }

}